replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/HeartbeaterAndVerifierImpl.java

```diff
@@ -73,23 +135,307 @@ public void cleanUp() {
+
+    systemSubject.execute(() -> processSiteAsAdmin(site));
+  }
+
+  /**
+   * If the heartbeat fails, get all configurations that involve that site and mark their status to
+   * Status.CONNECTION_LOST if the current status was XXXX_IN_PROGRESS and to
+   * Status.CONNECTION_UNAVAILABLE for any other status
+   *
+   * <p>If the heartbeat succeeds, get all configurations that involves that site and if their
+   * current status is Status.CONNECTION_LOST or Status.CONNECTION_UNAVAILABLE, mark their status to
+   * Status.PENDING indicating the remote end should restart soon to replicate. For all other
+   * status, leave untouched
+   */
+  private void processSiteAsAdmin(ReplicationSite site) {
+    LOGGER.trace("Processing heartbeat for {} with subject", site.getUrl());
+    verify(site);
+    // next let's make sure it is a remotely managed site
+    if (!site.isRemoteManaged()) {
+      LOGGER.trace("Site {}'s is not remotely managed, not sending a heartbeat", site.getUrl());
+    } else if (site.getVerifiedUrl() == null) {
+      changeConfigStatusFor(
+          site,
+          (replicationStatus, status) -> {
+            if (status == Status.PULL_IN_PROGRESS || status == Status.PUSH_IN_PROGRESS) {
+              // todo do we need to cancel the running job as well? ie push/pull in progress
+              // status will just be overridden for current running jobs
+              replicationStatus.setStatus(Status.CONNECTION_LOST);
+            } else {
+              replicationStatus.setStatus(Status.CONNECTION_UNAVAILABLE);
+            }
+          });
+    } else {
+      changeConfigStatusFor(
+          site,
+          (replicationStatus, status) -> {
+            if (status == Status.CONNECTION_LOST || status == Status.CONNECTION_UNAVAILABLE) {
+              replicationStatus.setStatus(Status.PENDING);
+            }
+          });
+    }
+  }
+
+  private void changeConfigStatusFor(
+      ReplicationSite site, BiConsumer<ReplicationStatus, Status> updater) {
+    configManager
+        .objects()
+        .filter(config -> config.sourceOrDestinationIs(site.getId()))
+        .forEach(
+            config -> {
+              Optional<ReplicationStatus> statusOptional = statusFor(config);
+              if (statusOptional.isPresent()) {
+                ReplicationStatus replicationStatus = statusOptional.get();
+                updater.accept(replicationStatus, replicationStatus.getStatus());
+
+                try {
+                  // todo this is going to undesirably condense the status' stats...
+                  history.addReplicationEvent(replicationStatus);
+                } catch (ReplicationPersistenceException e) {
+                  LOGGER.debug(
+                      "Failed to update status while heartbeating site {}", site.getUrl(), e);
+                }
+              } else {
+                LOGGER.warn("No existing status to update for config {}", config.getName());
```

Comment:
❓ I think we need to create one otherwise the first time around we are trying to verify a site will never be reported as CONNECTION_UNAVAILABLE.